cirq/work/observable_measurement.py

```diff
@@ -0,0 +1,1317 @@
+    and has the maximum non-identity elements.
+
+    The returned PauliString is constructed by taking the non-identity
+    single-qubit Pauli at each qubit position.
+
+    This function will return `None` if the input observables do not share a
+    tensor product basis.
+
+    For example, the _max_weight_observable of ["XI", "IZ"] is "XZ". Asking for
+    the max weight observable of something like ["XI", "ZI"] will return None.
+
+    The returned value need not actually be present in the input observables.
+    Coefficients from input observables will be dropped.
+    """
+    qubit_pauli_map = dict()  # type: Dict[ops.Qid, ops.Pauli]
+    for observable in observables:
+        for qubit, pauli in observable.items():
+            if qubit in qubit_pauli_map:
+                if qubit_pauli_map[qubit] != pauli:
+                    return None
+            else:
+                qubit_pauli_map[qubit] = pauli
+    return ops.PauliString(qubit_pauli_map)
+
+
+def _max_weight_observable_strict(
+        pauli_strings: Iterable[ops.PauliString]) -> ops.PauliString:
+    """Like `_max_weight_operator`, but will raise a ValueError instead of
+    returning None for non-compatible observables.
+    """
+    max_weight_op = _max_weight_observable(pauli_strings)
+    if max_weight_op is None:
+        raise ValueError("Grouping error for operators {}"
+                         .format(pauli_strings))
+    return max_weight_op
+
+
+def _max_weight_state(states: Iterable[value.TensorProductState]) \
+        -> Union[None, value.TensorProductState]:
+    """Create a new state that is compatible with all input states
+    and has the maximum weight.
+
+    The returned TensorProductState is constructed by taking the
+    single-qubit state at each qubit position.
+
+    This function will return `None` if the input states are not compatible
+
+    For example, the max_weight_state of [+X(0), -Z(1)] is
+    "+X(0) * -Z(1)". Asking for the max weight state of something like
+    [+X(0), +Z(0)] will return None.
+    """
+    qubit_state_map = dict()  # type: Dict[ops.Qid, _Named1QState]
+    for state in states:
+        for qubit, named_state in state:
+            if qubit in qubit_state_map:
+                if qubit_state_map[qubit] != named_state:
+                    return None
+            else:
+                qubit_state_map[qubit] = named_state
+    return value.TensorProductState(qubit_state_map)
+
+
+def _max_weight_state_strict(states: Iterable[value.TensorProductState]) \
+        -> value.TensorProductState:
+    """Like `_max_weight_state`, but will raise a ValueError instead of
+    returning None for non-compatible states.
+    """
+    max_weight_state = _max_weight_state(states)
+    if max_weight_state is None:
+        raise ValueError("Grouping error for states {}".format(states))
+    return max_weight_state
+
+
+def group_settings_greedy(settings: Iterable[InitObsSetting]) \
```

Comment:
I think the current design is not generic enough to enable different types of grouping strategies because:

- Different types of grouping strategies can introduce different unitary rotation gates at the end of the circuit for each group. 
- The post-processing for obtaining the results for individual `List[InitObsSetting]` from the single `InitObsSetting` can be different for every grouping strategy. 
- The final measurement strategy can be different for each grouping strategy. For eg: basis change + measuring a single qubit in Z basis (using `PauliMeasurementGate` proposed in   https://github.com/quantumlib/Cirq/issues/942#issuecomment-658933767) instead of measuring all qubits in the Z basis (current `cirq.measure(qubits)`) (I don't have an immediate example for this).

An example for the first two cases would be if we wish to implement the grouping strategies proposed in [arxiv:1907.09386](https://arxiv.org/pdf/1907.09386.pdf) or [arxiv:1908.06942](https://arxiv.org/pdf/1908.06942.pdf)

In order to fix these issues, I propose that we have a separate base class for defining the grouping strategy which performs the grouping, returns a `circuit_to_append` including measurement gates and consumes the results of measurement to perform the post-processing. The expectation corresponding to every original `InitObsSetting` can then be queried from this.

Reply:
At the risk of making this PR even more untenably large, can we punt on more general term measurement strategies? Your proposal makes sense at a surface level but I think it would work better as a follow-up change with a fully-featured implementation of 1907.09386 instead of providing a set of api hooks we *think* will be sufficient for the follow on feature.

What do you think?

Reply:
The primary concern I have with your proposal is that the PR implementing 1907.09386 would also be responsible to refactor the work done here OR leave it in a legacy state (which is bad). The refactoring would also deprecate the current `measure_observables` interface, which will be an added effort. 
Let's continue the discussion in cirq sync.